src/main/java/Controller/spaceController.java

```diff
@@ -1,41 +1,100 @@
-        this.spaceAddInteractor = spaceAddInteractor;
-        this.spacePlayInteractor = spacePlayInteractor;
+    public spaceController(playSpaceInputBoundary spacePlayInteractor, spacePresenter spacePresenter){
+        this.playSpaceInteractor = spacePlayInteractor;
+        this.spaceSongList = new ArrayList<>();
+        this.sync = MusicPlayer.getInstance().getSync();
+        this.keepPlaying = true;
+        this.spacePresenter = spacePresenter;
     }
 
     /**
      * function calling the use case for playing the space
-     * @param spaceInputData input data for playing the space
      */
-    public void playSpace(spaceInputData spaceInputData){
-        this.spacePlayInteractor.playSpace(spaceInputData);
+    public void playSpace(){
+        // pick which song to be played
+        songOutputData songToPlay;
+        if (!this.spaceSongList.isEmpty()){
+            songToPlay = new songOutputData(this.spaceSongList.remove(0).getSong());  // remove song to be played
+        }
+        else{
+            songToPlay = this.playSpaceInteractor.pickRandomSong();
+        }
+
+        // open queue
+        final Thread thread = new Thread(this::playNextSong);
+        thread.start();
+
+        // call presenter
+        this.spacePresenter.spacePlayed();  // update button
+        this.spacePresenter.songPlayed(songToPlay);  // update playbar
+
+        // play the song
+        this.playSpaceInteractor.playSpace(new songInputData(songToPlay.getSong()));
+    }
+
+    /**
+     * helper function for playSpace()
+     */
+    private void playNextSong(){
+        synchronized (sync){
+            try{
+                sync.wait();
+            } catch (InterruptedException e) {
+                System.out.println("thread interrupted");
+            }
+            if (keepPlaying){  // start all over again
+                this.playSpace();
+            }
+        }
+    }
+
+    /**
+     * in order to stop infinite recursion, call stopSpace.
+     */
+    private void stopSpace(){
+        this.keepPlaying = false;
     }
 
     /**
      * function calling the use case for adding a song to the space
-     * @param spaceInputData input data for adding a song to the space
+     * @param songInputData input data for adding a song to the space
      */
-    public void spaceAddSong(spaceInputData spaceInputData){
-        this.spaceAddInteractor.spaceAddSong(spaceInputData);
+    public void spaceAddSong(songInputData songInputData){
+        Integer songToAddID = songInputData.getSong().getID();
+        for (songInputData currSongInputData : this.spaceSongList){
+            Integer currSongID = currSongInputData.getSong().getID();
+            if (currSongID.equals(songToAddID)){
+                return;  // if the song is already in the playlist, do nothing
+                // TODO: make this cooler (i.e. upvote algo) if time
```

Comment:
If you do implement the upvote algorithm, you might want to keep in mind it wouldn't _just_ be that -- you'd have to have the GUI that comes with it, which would either 1) significantly affect the layout of the dashboard or 2) give space it's own separate page that shows everything in the user-requested queue. (likely the latter). 

tldr this is less of a 'if we have time' feature and more of a 'plan in advance' one

Reply:
I was initially thinking of instead having the upvote process itself on the GUI, it'd just be handled within that function -- i.e. make it so that when you add a song to the space that is already queued, it might update the song's position within the queue, but there's no explicit "upvote" button. I could conceivably see changing the output to the presenter (if indeed we're doing that -- last I remember, add to space doesn't have an output data, @SiqiZhang6 could you confirm?) to say something like "song recommended to be played sooner!" instead of "song added to space!" or something, but nothing major.

Reply:
>make it so that when you add a song to the space that is already queued, it might update the song's position within the queue, but there's no explicit "upvote" button. 
If it's just that, I don't see a reason why we couldn't just do it now, lol.

Reply:
I mean we probably could, I just haven't figured out what that will look like in actual code, lol. Like what determines an increase in 1 position? At what point do you stop increasing (assuming we want the thing where we weigh both time in the queue + popularity)? etc. Once that's settled, I think actually implementing it shouldn't be that bad.

Reply:
> last I remember, add to space doesn't have an output data, @SiqiZhang6 could you confirm?

yes there is no output data, unless you wanna add one